indra/llwindow/llgamecontrol.cpp

```diff
@@ -585,28 +851,54 @@ void LLGameControllerManager::initializeMappingsByDefault()
     };
 }
 
-void LLGameControllerManager::addController(SDL_JoystickID id, SDL_GameController* controller)
+void LLGameControllerManager::resetDeviceOptionsToDefaults()
+{
+    for (LLGameControl::Device& device : mDevices)
+    {
+        device.resetOptionsToDefaults();
+    }
+}
+
+void LLGameControllerManager::loadDeviceOptionsFromSettings()
 {
-    LL_INFOS("GameController") << "joystick id: " << id << ", controller: " << controller << LL_ENDL;
+    for (LLGameControl::Device& device : mDevices)
+    {
+        device.loadOptionsFromString(g_deviceOptions[device.getGUID()]);
+    }
+}
+
+void LLGameControllerManager::saveDeviceOptionsToSettings() const
+{
+    for (const LLGameControl::Device& device : mDevices)
+    {
+        std::string options = device.saveOptionsToString();
+        if (options.empty())
+        {
+            g_deviceOptions.erase(device.getGUID());
+        }
+        else
+        {
+            g_deviceOptions[device.getGUID()] = options;
+        }
+    }
+}
 
+void LLGameControllerManager::addController(SDL_JoystickID id, const std::string& guid, const std::string& name)
+{
     llassert(id >= 0);
-    llassert(controller);
 
     if (findDevice(id) != mDevices.end())
     {
-        LL_WARNS("GameController") << "device already added" << LL_ENDL;
+        LL_WARNS("SDL2") << "device id=" << id << " was already added" << LL_ENDL;
```

Comment:
I suggest reporting guid and name here just in case. I'm not sure how those ids work, but I doubt id is helpfull for figuring out what happened.

Reply:
This logging is being done immediately before the call `addController`